### ISSUE TITLE
Translated strings don't wait for page transition

### ISSUE THREAD

**Issue by amsousa97** [nuxt-modules/i18n#1255]
### Version
@nuxtjs/i18n: 7.0.1
nuxt: 2.15.7

### Nuxt configuration

Please change to `[x]` if relevant for this issue:

 - [ ] Applies to a site deployed to a static server (site generated with `nuxt generate`)
 - [X] Applies to a site deployed to a server with a Node backend

### @nuxtjs/i18n configuration
```
i18n: {
    locales: [
      { code: "en", iso: "en-US", name: "English", file: "en-US" },
      { code: "pt", iso: "pt-PT", name: "Português", file: "pt-PT" }
    ],
    defaultLocale: "en",
    detectBrowserLanguage: {
      redirectOn: "root"
    },
    strategy: "prefix",
    lazy: true,
    langDir: "i18n/",
    parsePages: false,
    vueI18n: {
      fallbackLocale: "en"
    },
    skipSettingLocaleOnNavigate: true
  }
```

### Reproduction Link
I created a Code Sandbox: https://codesandbox.io/s/cocky-galileo-z8qhk?file=/pages/index.vue

### Steps to reproduce
Access the code sandbox and click either one of the language selection links.

### What is Expected?
The translated string should only change after the page transition completed.

### What is actually happening?
The translated string changes before transition is completed and is then re-rendered.

**Comment by rchl** [comment]
That logic doesn't work when explicitly using `$i18n.setLocale` which you are doing from `LangSwitcher.vue`. It only works on navigating to a route with a different locale.

You'd want to change the lang switcher to something like:

```vue
    <nuxt-link
      v-for="locale of $i18n.locales"
      :key="locale.code"
      :to="switchLocalePath(locale.code)"
    >
      {{ locale.name }}
    </nuxt-link>
```

**Comment by rchl** [comment]
I guess that it could still be seen as a bug that it doesn't work using `setLocale`.

**Comment by amsousa97** [comment]
Thanks, your suggestion worked! 
However, on my app, whenever I change the locale I have to perform a request to get the data for the matching locale.
With my previous approach was using ```i18n.locale``` inside ```asyncData``` to get the data for that locale.
This seems to break when using ```switchLocalePath``` and ```nuxt-link```, so I'm assuming I'd have to change to ```fetch``` instead?

**Comment by rchl** [comment]
The `asyncData` should still be called in that case as you are changing to a different route.

I've tried with:

```vue
<template>
  <div>{{ $t("welcome") }}</div>
</template>

<script>
export default {
  asyncData({ route }) {
    console.log(`asyncData for route ${route.path}`)
  },
}
</script>
```

And the log was printed on changing locale.

**Comment by amsousa97** [comment]
After fiddling around with this it appears that while ```asyncData``` is called correctly, the ```locale``` property seems to be outdated. So when I do something like this (this is from the docs)

```javascript
export default Vue.extend({
  asyncData({ i18n }) {
    const locale = i18n.locale

    return {
      locale
    }
  }
})
```
the old locale would be set and not the that's currently being switched to. The only I'm currently able to get the pending locale is through the ```__pendingLocale``` property, but it doesn't seem like a good idea to use that.

**Comment by rchl** [comment]
Hmm... we want the locale to change only after the new route has loaded but also we want the locale property to be updated before the new route starts loading... Those are conflicting expectations.

I'm not sure if there is a clean solution for that.

**Comment by rchl** [comment]
Will re-open although I don't have any possible solution in mind right now.

**Comment by vampics** [comment]
+1

**Comment by lumo10** [comment]
+1

**Comment by gerbenvandijk** [comment]
This is an issue for me as well; the documented solution (`skipSettingLocaleOnNavigate: true` in `i18n` object in `nuxt.config.ts`) does not work when doing e.g. `navigateTo(switchLocalePath(locale))`.

I've confirmed the documented solution partially "works" with navigating via `<nuxt-link :to="navigateTo(switchLocalePath(locale)"...` but it would be great to provide full support for other methods of changing route in Nuxt. 

In my case I am using a `<select @change="navigateTo(switchLocalePath(locale))">...` element so that users on mobile devices can switch language in an intuitive way and we don't have to build a custom dropdown element that works on all screensizes etc. 

I am saying "partly works" is because setting `skipSettingLocaleOnNavigate: true` in `nuxt.config.ts` does prevent locale changing when using <nuxt-link>. However the documented solution for switching locale before the enter transition via the `defaultTransition.beforeEnter` navigation guard does **not** work  (e.g. `Cannot read properties of undefined (reading 'defaultTransition')`. 

I've tried multiple ways of setting this via nuxt.config.ts app.pageTransition. beforeEnter and also via router.options.ts (where according to Nuxt documentation you should do these type of things), but in both cases there is no `nuxtApp` context available. The only thing that seems to "work" is the above. 

For the people who can rely on using <nuxt-link> for this you can find a working solution below:

- `skipSettingLocaleOnNavigate: true` in `nuxt.config.ts` -> `i18n` object
- create a client side only plugin with the code below:
```ts
export default defineNuxtPlugin(nuxtApp => {
    const router = useRouter()
    router.afterEach(() => {
        setTimeout(() => {
            nuxtApp.$i18n.finalizePendingLocaleChange()
        }, 300) // whatever the duration of your page transition is
    })
})
```

Would be good if a good solution could be provided by the package itself though, switching language looks quite bugged out now by default and it should not be needed to write all this code for such a basic thing.

**Comment by gerbenvandijk** [comment]
For the people who can rely on using <nuxt-link> for this you can find a working solution below:

- `skipSettingLocaleOnNavigate: true` in `nuxt.config.ts` -> `i18n` object
- create a plugin with the code below:
```ts
export default defineNuxtPlugin(nuxtApp => {
    const router = useRouter()
    router.afterEach(() => {
        setTimeout(() => {
            nuxtApp.$i18n.finalizePendingLocaleChange()
        }, 300) // whatever the duration of your page transition is
    })
})
```

**Comment by notflip** [comment]
@gerbenvandijk I just did a fresh install and the default setup from the docs: https://i18n.nuxtjs.org/docs/guide/lang-switcher#wait-for-page-transition seems to work, can you confirm it's fixed? I'm using

```
"nuxt": "^3.10.1",
"@nuxtjs/i18n": "^8.1.0",
```

This is my i18n config

```
    i18n: {
        skipSettingLocaleOnNavigate: true,
        strategy: 'prefix_except_default',
        defaultLocale: 'nl',
        langDir: 'lang',
        locales: [
            {code: 'nl', file: 'nl.json', iso: 'nl', name: 'NL'},
            {code: 'en', file: 'en.json', iso: 'en', name: 'EN'},
        ],
        detectBrowserLanguage: false,
        customRoutes: "page",
    },
```

**Comment by gerbenvandijk** [comment]
Will have a look at it on monday and let you know! Thanks for looking at it :)

**Comment by notflip** [comment]
> Will have a look at it on monday and let you know! Thanks for looking at it :)

Never mind! It doesn't work when I'm on a detail page such as `/en/startup`, when I use switchLocale it calls my api with `{locale: nl, slug: startup}`, it should be `{locale: nl, slug: opstarten}`. Sorry for the false hope.

Your proposed solution is working though, so thank you.

**Comment by BobbieGoede** [comment]
@notflip @gerbenvandijk 
FYI, this issue is regarding v7 (Nuxt 2), if you're experiencing an issue with v8 please open a new issue with a (minimal) reproduction!

**Comment by ghost** [comment]
Same problem here. Looking for a fix.

**Comment by BobbieGoede** [comment]
Closing this as v7 is not being actively worked on (critical hotfixes only), if this still applies to v8 and later please open a new issue!
